### DISCUSSION TITLE
Independent Node.js Headers module

### DISCUSSION THREAD

**Discussion by Ethan-Arrowood** [Ethan-Arrowood/undici-fetch#35] [Ideas]
https://github.com/Ethan-Arrowood/undici-fetch/pull/32 has sparked an interesting discussion about Headers in Node.js

I'm very open to the idea of a standalone headers core module for Node.js. We could contribute this directly to Node core or it can exist similar to nodejs/undici repo for now. This standardization would be very helpful for libraries such as undici, undici-fetch, and many others.

WDYT?

**Comment by Ethan-Arrowood** [comment]
@mcollina @timdp FYI let's discuss headers stuff more here

**Comment by Ethan-Arrowood** [comment]
One important aspect of a headers module would be that it standardizes how headers are handled across node ecosystem.

Any sort of Headers class will never beat the performance of just a regular object, but I think we could create something as-performant-as-possible and then also provide utilities that allow for even faster use cases.

One such example I'm thinking of is if a module wants to manage headers using an array. We could provide utilities that operate specifically for arrays; or we at least provide the core APIs necessary to do so. Normalization, validation, joining, etc.

**Reply by timdp** [reply]
People also use arrays because the same header name can occur multiple times, and that's tedious to solve with an object.

I'm also not overly concerned about performance to be honest. Given the order of magnitude, even a Headers class backed by a naive array would probably be fine, even if you regularly perform name lookups. As I understand it, that's how small objects work underneath as well.

I'm not familiar with the quirks of the Fetch API's Headers though. Anything in particular that's slowing it down? Weird preprocessing or so?

**Reply by ronag** [reply]
Using an array as internal store is an interesting idea. Would make some things faster. I think it makes sense to optimize for fast append + iteration over fast random lookups.

**Reply by ronag** [reply]
Maybe such an implementation could even be faster than the POJO based approach? @Ethan-Arrowood wdyt?

**Reply by timdp** [reply]
You could also keep it sorted by header name and use binary search for lookup. Not sure if it's worth the effort though.

**Reply by ronag** [reply]
That would slow down append.

**Reply by timdp** [reply]
Well, yeah, it's going to depend on the use case.

**Reply by Ethan-Arrowood** [reply]
Here is my first pass on an array based headers class

```js
class ArrayHeaders {
  constructor (init) {
    this[kHeaders] = []

    if (init && typeof init === 'object') {
      fill(this, init)
    }
  }

  append (name, value) {
    const newHeader = normalizeAndValidateHeaderValue(name, value)

    let isNewEntry = true
    for (var i = 0; i < this[kHeaders].length; i++) {
      if (newHeader[0] === this[kHeaders][i][0]) {
        this[kHeaders][i][1] = `${this[kHeaders][i][1]}, ${newHeader[1]}`
        isNewEntry = false
        break
      }
    }

    if (isNewEntry) {
      this[kHeaders].push(newHeader)
    }
  }

  delete (name) {
    const normalizedHeaderName = normalizeAndValidateHeaderName(name)

    for (var i = 0; i < this[kHeaders].length; i++) {
      if (normalizedHeaderName === this[kHeaders][i][0]) {
        this[kHeaders].splice(i, 1)
        break
      }
    }
  }

  get (name) {
    const normalizedHeaderName = normalizeAndValidateHeaderName(name)

    for (var i = 0; i < this[kHeaders].length; i++) {
      if (normalizedHeaderName === this[kHeaders][i][0]) {
        return this[kHeaders][i][1] ?? null
      }
    }
  }

  has (name) {
    const normalizedHeaderName = normalizeAndValidateHeaderName(name)

    for (var i = 0; i < this[kHeaders].length; i++) {
      if (normalizedHeaderName === this[kHeaders][i][0]) {
        return true
      }
    }

    return false
  }

  set (name, value) {
    const newHeader = normalizeAndValidateHeaderValue(name, value)

    let isNewEntry = true
    for (var i = 0; i < this[kHeaders].length; i++) {
      if (newHeader[0] === this[kHeaders][i][0]) {
        this[kHeaders][i] = newHeader
        isNewEntry = false
        break
      }
    }

    if (isNewEntry) {
      this[kHeaders].push(newHeader)
    }
  }

  * keys () {
    this[kHeaders].sort()
    for (var i = 0; i < this[kHeaders].length; i++) {
      yield this[kHeaders][i][0]
    }
  }

  * values () {
    this[kHeaders].sort()
    for (var i = 0; i < this[kHeaders].length; i++) {
      yield this[kHeaders][i][1]
    }
  }

  * entries () {
    yield * this[kHeaders].sort()
  }

  forEach (callback, thisArg) {
    for (const [key, value] of this) {
      callback.call(thisArg, value, key, this)
    }
  }

  [Symbol.iterator] () {
    return this.entries()
  }
}
```

This implementation is built on a 2D array of header name/value pairs. 

> You could also keep it sorted by header name and use binary search for lookup. Not sure if it's worth the effort though.

I was thinking exactly this in order to keep the iteration performant. But don't have the effort currently to implement that. I'm surprised `.sort` works on an array of tuples.

**Reply by timdp** [reply]
Sorting probably works because it's stringifying first, but that's wildly inefficient.

Does the spec say they need to be sorted?

Also, an array of tuples is going to allocate a lot of refs. As I said elsewhere, I think that's why Node uses a zipped 1D array.

Also², I think you wrote var where you meant let.

And this should probably be a PR?

**Reply by Ethan-Arrowood** [reply]
> Sorting probably works because it's stringifying first, but that's wildly inefficient.

Oh good point

> Does the spec say they need to be sorted?

Yes the iterator must return a sorted and combined list of headers

> Also, an array of tuples is going to allocate a lot of refs. As I said elsewhere, I think that's why Node uses a zipped 1D array.

True, shouldn't be too difficult to change this to a zipped 1D array

> Also², I think you wrote var where you meant let.

Ah yep whoops

> And this should probably be a PR?

🤷‍♂️ wasn't sure the best place to sandbox this. I can make a new repo where we can checking and test various implementations if we want

**Reply by ronag** [reply]
I was more thinking of a 1D array rather than 2D. You kind of lose the efficiency with a 2D array.

**Reply by ronag** [reply]
We would need some way to access the non-sorted raw array or the performance boost is lost.

**Reply by Ethan-Arrowood** [reply]
we can export the symbol so those that know what they are doing can get the direct access

**Reply by Ethan-Arrowood** [reply]
Im working on a 1D version now. Sorting is much more difficult.

**Reply by Ethan-Arrowood** [reply]
```js
class ArrayHeaders {
  constructor (init) {
    this[kHeaders] = []

    if (init && typeof init === 'object') {
      fill(this, init)
    }
  }

  append (name, value) {
    const [normalizedName, normalizedValue] = normalizeAndValidateHeaderValue(name, value)

    let isNewEntry = true
    for (var i = 0; i < this[kHeaders].length; i+=2) {
      if (normalizedName === this[kHeaders][i]) {
        this[kHeaders][i+1] += `, ${normalizedValue}`
        isNewEntry = false
        break
      }
    }

    if (isNewEntry) {
      this[kHeaders].push(normalizedName)
      this[kHeaders].push(normalizedValue)
    }
  }

  delete (name) {
    const normalizedName = normalizeAndValidateHeaderName(name)

    for (var i = 0; i < this[kHeaders].length; i+=2) {
      if (normalizedName === this[kHeaders][i]) {
        this[kHeaders].splice(i, 2)
        break
      }
    }
  }

  get (name) {
    const normalizedName = normalizeAndValidateHeaderName(name)

    for (var i = 0; i < this[kHeaders].length; i+=2) {
      if (normalizedName === this[kHeaders][i]) {
        return this[kHeaders][i+1]
      }
    }

    return null
  }

  has (name) {
    const normalizedName = normalizeAndValidateHeaderName(name)

    for (var i = 0; i < this[kHeaders].length; i+=2) {
      if (normalizedName === this[kHeaders][i]) {
        return true
      }
    }

    return false
  }

  set (name, value) {
    const [normalizedName, normalizedValue] = normalizeAndValidateHeaderValue(name, value)

    let isNewEntry = true
    for (var i = 0; i < this[kHeaders].length; i+=2) {
      if (normalizedName === this[kHeaders][i]) {
        this[kHeaders][i+1] = normalizedValue
        isNewEntry = false
        break
      }
    }

    if (isNewEntry) {
      this[kHeaders].push(normalizedName)
      this[kHeaders].push(normalizedValue)
    }
  }

  * keys () {
    for (const header of this) {
      yield header[0]
    }
  }

  * values () {
    for (const header of this) {
      yield header[1]
    }
  }

  * entries () {
    yield * this
  }

  forEach (callback, thisArg) {
    this[kHeaders] = sort(this[kHeaders])
    for (let i = 0; i < this[kHeaders].length; i+=2) {
      callback.call(thisArg, this[kHeaders][i+1], this[kHeaders][i], this)
    }
  }

  * [Symbol.iterator] () {
    this[kHeaders] = sort(this[kHeaders])
    for (let i = 0; i < this[kHeaders].length; i+=2) {
      yield [this[kHeaders][i], this[kHeaders][i+1]]
    }
  }
}

function sort (headers) {
  const names = []
  for (let i = 0; i < headers.length; i+=2) {
    names.push(headers[i])
  }
  names.sort()
  const sorted = []
  for (const name of names) {
    sorted.push(name)
    sorted.push(headers[headers.indexOf(name) + 1])
  }
  return sorted
}
```

Sorting algorithm is not optimized at all; literally just wrote the first thing that came to mind. I haven't attempted to benchmark any of this yet, but at least this is using a single internal array

**Comment by Ethan-Arrowood** [comment]
Writing an implementation using V8 internals or compiling to WASM is a great idea. Llhttp has proven that is a great solution for generally non-performant Web compatibility things

**Reply by Ethan-Arrowood** [reply]
FWIW I'm very interested in writing my first WASM module. I've been learning Rust in my free time if we could try to use that; otherwise I'm happy to use C++ too

**Reply by mcollina** [reply]
Rust would be great!

**Reply by timdp** [reply]
Would it though? I expect overhead from serialization, fiddling with shared memory, and a huge barrier to entry from people looking to contribute. All to shave off a millisecond of processing for an HTTP client?

**Reply by mcollina** [reply]
I have no idea if it would be worth it. The majority of the costs in managing headers come from the case insensitivity and all the arrays/objects needed to support that. If we could have a data structure where we put headers in and we get a string ready to be sent on the wire out... it might be well worth it. Some research on fast data structures might be necessary.

Regarding the 1ms comment, I'll try to explain my point of view. Let's consider a deployment where we have 1 CPU and we would like to handle 1000 HTTP req/sec. The max CPU processing time for each request is 1ms. If our target became 10k req/sec, then our budget is 0.1ms.

The goal for a utility library to make HTTP requests is to not be the bottleneck - i.e. optimizations in that library are unlikely to be measurable by the users. This usually means that the CPU time spent in frameworks and libraries must be one or two order of lower than the time spent in the business logic.

**Reply by timdp** [reply]
I'm not advocating that we dismiss the idea without further research but let me illustrate our use case.

For many years, we've been operating a proxy service that pulls in XML over HTTP, rewrites it, and sends it back to the client. Engineering for performance has been a challenge.

Obviously, we've had to optimize the heck out of the XML processing. When we got that under control, we were able to focus more on HTTP requests.

Frankly, based on the profiler, header processing was never on our radar. Instead, we've been investing our time in things like DNS caching, pipelining, and h2 support (and hopefully QUIC at some point). All of those are fairly critical, yet still tricky to get right today. If it were up to me, I'd make sure a new HTTP client was able to address those needs, more than trying to save nanoseconds on headers.

I realize that a) I'm biased, b) these are nontrivial concerns, and c) they're not mutually exclusive. However, in reality, I do think priorities are going to matter a lot.

**Reply by mcollina** [reply]
The nicest thing about Open Source is that everybody can volunteer to work on what matters to them. Nobody is preventing you to contribute wherever you see fit and a maintainer is happy to land your changes (or give you the commit bit). This is definitely true on Undici and Node.js.

All those things you mentioned are important, there is really a lot to work on!

**Reply by timdp** [reply]
That's one of the reasons why I started to contribute to undici-fetch. But it doesn't mean that premature optimization is a good idea. 🙂

**Comment by Ethan-Arrowood** [comment]
Copying pasting relevant comment thread from a recent Headers related PR that has some additional ideas: https://github.com/Ethan-Arrowood/undici-fetch/pull/36#discussion_r607375622

**Comment by Ethan-Arrowood** [comment]
https://github.com/nodejs/node/pull/38986
